disperser/controller/dispatcher.go

```diff
@@ -0,0 +1,432 @@
+	currentBlockNumber, err := d.chainState.GetCurrentBlockNumber()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get current block number: %w", err)
+	}
+	referenceBlockNumber := uint64(currentBlockNumber) - d.FinalizationBlockDelay
+
+	// Get a batch of blobs to dispatch
+	// This also writes a batch header and blob verification info for each blob in metadata store
+	batchData, err := d.NewBatch(ctx, referenceBlockNumber)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	batch := batchData.Batch
+	state := batchData.OperatorState
+	sigChan := make(chan core.SigningMessage, len(state.IndexedOperators))
+	for opID, op := range state.IndexedOperators {
+		opID := opID
+		op := op
+		client, err := d.nodeClientManager.GetClient(opID, op.Socket)
+		if err != nil {
+			d.logger.Error("failed to get node client", "operator", opID, "err", err)
+			continue
+		}
+
+		d.pool.Submit(func() {
+
+			req := &corev2.DispersalRequest{
+				OperatorID: opID,
+				// TODO: get OperatorAddress
+				OperatorAddress: gethcommon.Address{},
+				Socket:          op.Socket,
+				DispersedAt:     uint64(time.Now().UnixNano()),
+				BatchHeader:     *batch.BatchHeader,
+			}
+			err := d.blobMetadataStore.PutDispersalRequest(ctx, req)
+			if err != nil {
+				d.logger.Error("failed to put dispersal request", "err", err)
+				return
+			}
+
+			for i := 0; i < d.NumRequestRetries+1; i++ {
+				sig, err := d.sendChunks(ctx, client, batch)
+				if err == nil {
+					storeErr := d.blobMetadataStore.PutDispersalResponse(ctx, &corev2.DispersalResponse{
+						DispersalRequest: req,
+						RespondedAt:      uint64(time.Now().UnixNano()),
+						Signature:        sig.Bytes(),
+						Error:            "",
+					})
+					if storeErr != nil {
+						d.logger.Error("failed to put dispersal response", "err", storeErr)
+					}
+
+					sigChan <- core.SigningMessage{
+						Signature:            sig,
+						Operator:             opID,
+						BatchHeaderHash:      batchData.BatchHeaderHash,
+						AttestationLatencyMs: 0, // TODO: calculate latency
+						Err:                  nil,
+					}
+
+					break
+				}
+
+				d.logger.Warn("failed to send chunks", "operator", opID, "NumAttempts", i, "err", err)
+				time.Sleep(time.Duration(math.Pow(2, float64(i))) * time.Second) // Wait before retrying
+			}
+		})
+	}
+
+	return sigChan, batchData, nil
+}
+
+// HandleSignatures receives signatures from operators, validates, and aggregates them
+func (d *Dispatcher) HandleSignatures(ctx context.Context, batchData *batchData, sigChan chan core.SigningMessage) error {
+	quorumAttestation, err := d.aggregator.ReceiveSignatures(ctx, batchData.OperatorState, batchData.BatchHeaderHash, sigChan)
+	if err != nil {
+		return fmt.Errorf("failed to receive and validate signatures: %w", err)
+	}
+	numPassed, passedQuorums := numBlobsAttestedByQuorum(quorumAttestation.QuorumResults, batchData.Batch.BlobCertificates)
```

Comment:
Is deduplication of certs already handled upstream ?

Reply:
There is no deduplication. New batch should be constructed by blobs that haven't been processed yet. `GetBlobMetadataByStatus` takes `lastUpdatedAt` which gets updated as we pull new blobs and acts as a cursor. It only takes blobs that are updated after this value. 

`numBlobsAttestedByQuorum` is added to 1) early terminate if no blobs received sufficient amount of signatures, or 2) remove unsuccessful quorums (to help reduce gas costs for bridging). However, this optimization seems pretty minor, and I think I'll remove this check to simplify.